parsons/mobilecommons/mobilecommons.py

```diff
@@ -0,0 +1,122 @@
+                The endpoint, which will be appended to the base URL for each request
+            data_key: str
+                The key used to extract the desired data from the response dictionary derived from
+                the xml response
+            params: str
+                Parameters to be passed into GET request
+            elements_to_unpack: list
+                A list of elements that contain dictionaries to be unpacked into new columns in the
+                final table
+            limit: int
+                The maximum number of rows to return
+        `Returns:`
+            Parsons table with requested data
+        """
+
+        # Create a table to compile results from different pages in
+        final_table = Table()
+        # Max page_limit is 1000 for MC
+        page_limit = 1000 if limit > 1000 or limit is None else limit
+        params += f'limit={str(page_limit)}'
+
+        logger.info(f'Working on fetching first {page_limit} rows. This can take a long time.'
+                    f'Each 1000 rows can take between 30-60 seconds to fetch.')
+
+        response = self.client.request(endpoint + self.companyid_param, 'GET', params=params)
+
+        # If good response, compile data into final_table
+        if response.status_code == 200:
+            # Parse xml to nested dictionary and load to parsons table
+            response_dict = xmltodict.parse(response.text, attr_prefix='', cdata_key='',
+                                            dict_constructor=dict)
+            response_table = Table(response_dict['response'][endpoint][data_key])
+            # Unpack any specified elements
+            for col in elements_to_unpack:
+                response_table.unpack_dict(col)
+            # Append to final table
+            final_table.concat(response_table)
+            # Check to see if there are more pages and determine how many to retrieve
+            avail_pages = int(response_dict['response'][endpoint]['page_count'])
+            req_pages = math.ceil(limit/page_limit)
+            pages_to_get = avail_pages if avail_pages < req_pages else req_pages
+            # Go fetch other pages of data
+            i = 2
+            while i <= pages_to_get:
+                page_params = params + f'&page={str(i)}'
+                logger.info(f'Fetching rows {str(i*page_limit)} - {str((i+1)*page_limit)} '
+                            f'of {limit}')
+                response = self.client.request(endpoint + self.companyid_param, 'GET',
+                                               params=page_params)
+                response_dict = xmltodict.parse(response.text, attr_prefix='', cdata_key='',
+                                                dict_constructor=dict)
+                response_table = Table(response_dict['response'][endpoint][data_key])
+                final_table.concat(response_table)
+                i += 1
+        # If initial response is bad, raise error
+        else:
```

Comment:
One other approach would be to check in if there's an error at the top and if not then fetch the data. It looks a little cleaner and keep you core logic from being indent more than it has to.

```python
if response.status_code != 200:
    # do error handling here
    ...
    raise

# do your data fetching here.
for i in range(2, pages_to_get + 1):
    ...
```